timeseries/src/autogluon/timeseries/utils/features.py

```diff
@@ -284,3 +296,63 @@ def _check_required_columns_are_present(
             raise ValueError(
                 f"{len(missing_columns)} columns are missing from {data_frame_name}: {reprlib.repr(missing_columns.to_list())}"
             )
+
+
+class AbstractFeatureImportanceTransform:
+    """Abstract class for transforms that replace a given feature with dummy or shuffled values,
+    for use in feature importance operations.
+    """
+
+    def __init__(
+        self,
+        forecast_target: str,
+        covariate_metadata: CovariateMetadata,
+    ):
+        self.forecast_target = forecast_target
+        self.covariate_metadata: CovariateMetadata = covariate_metadata
+
+    @classmethod
+    def from_feature_generator(
+        cls, feature_generator: TimeSeriesFeatureGenerator
+    ) -> "AbstractFeatureImportanceTransform":
+        return cls(
+            forecast_target=feature_generator.target,
+            covariate_metadata=feature_generator.covariate_metadata,
+        )
+
+    def _apply_transform(self, data: TimeSeriesDataFrame, target_feature: str) -> TimeSeriesDataFrame:
+        raise NotImplementedError
+
+    def transform(self, data: TimeSeriesDataFrame, target_feature: str, **kwargs) -> TimeSeriesDataFrame:
+        if target_feature not in self.covariate_metadata.all_features:
+            raise ValueError(f"Target feature {target_feature} not found in covariate metadata")
+
+        # feature transform works on a shallow copy of the main time series data frame
+        # but a deep copy of the static features.
+        data = data.copy(deep=False)
+
+        return self._apply_transform(data, target_feature, **kwargs)
+
+
+class PermutationFeatureImportanceTransform(AbstractFeatureImportanceTransform):
+    """Naively shuffles a given feature."""
+
+    def _apply_transform(self, data: TimeSeriesDataFrame, target_feature: str) -> TimeSeriesDataFrame:
+        if target_feature in self.covariate_metadata.static_features:
+            data.static_features[target_feature] = data.static_features[target_feature].sample(frac=1)
+        else:
+            data[target_feature] = data[target_feature].sample(frac=1)
+        return data
+
+
+class ConstantReplacementFeatureImportanceTransform(AbstractFeatureImportanceTransform):
+    """Replaces a target feature with the median if it's a real-valued feature, and the mode if it's a
+    categorical feature."""
+
+    def _apply_transform(self, data: TimeSeriesDataFrame, target_feature: str) -> TimeSeriesDataFrame:
+        aggregation = "median" if target_feature in self.covariate_metadata.real_features else "mode"
```

Comment:
I think it would be more prudent to use the mode/median of training data for masking here rather than mode/median of `data`. `data` will usually contain a subsample of 50 items from the full dataset, so the values used for imputation may change significantly based on the subset. 

Using statistics of the training data also has a minor advantage of being similar to our logic for imputing all-missing values.

Reply:
I'm not sure about 'training' data. I understand if it's not a subsample but the entire data provided during `feature_importance`. We do the imputations before shuffling, and I don't think there is an easy way of escaping

> so the values used for imputation may change significantly based on the subset